File: src/main/java/org/cimsbioko/utilities/LoginUtils.java
package org.cimsbioko.utilities;

import android.app.Activity;
import android.content.Intent;
import org.cimsbioko.activity.FieldWorkerLoginActivity;
import org.cimsbioko.model.core.FieldWorker;

import static android.content.Intent.FLAG_ACTIVITY_CLEAR_TASK;

public class LoginUtils {

    private static final Login<FieldWorker> login = new Login<>();

    public static class Login<U> {

        U authenticatedUser;

        private Login() {
        }

        public void setAuthenticatedUser(U user) {
            authenticatedUser = user;
        }

        public U getAuthenticatedUser() {
            return authenticatedUser;
        }

        public void logout(Activity ctx, boolean launchLoginActivity) {
            authenticatedUser = null;
            if (launchLoginActivity) {
                launchLogin(ctx);
            }
        }

    }

    public static Login<FieldWorker> getLogin() {
        return login;
    }

    static void launchLogin(Activity ctx) {
        ctx.startActivity(new Intent(ctx, FieldWorkerLoginActivity.class).setFlags(FLAG_ACTIVITY_CLEAR_TASK));
    }
}

File: src/main/java/org/cimsbioko/utilities/LoginUtils.kt
package org.cimsbioko.utilities

import android.app.Activity
import android.content.Intent
import org.cimsbioko.activity.FieldWorkerLoginActivity
import org.cimsbioko.model.core.FieldWorker

object LoginUtils {

    @JvmStatic
    val login = Login<FieldWorker>()

    @JvmStatic
    fun launchLogin(ctx: Activity) {
        ctx.startActivity(Intent(ctx, FieldWorkerLoginActivity::class.java).setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK))
    }

    class Login<U> internal constructor() {

        var authenticatedUser: U? = null

        fun logout(ctx: Activity, launchLoginActivity: Boolean) {
            authenticatedUser = null
            if (launchLoginActivity) {
                launchLogin(ctx)
            }
        }
    }
}